rust-src/concordium_base/src/curve_arithmetic/arkworks_instances.rs

```diff
@@ -0,0 +1,207 @@
+    const DOMAIN_STRING: &'static str;
+    type Hasher: ark_ec::hashing::HashToCurve<G>;
+}
+
+/// A blanket implementation of the `Curve` trait using the functionality of
+/// `ark_ec::CurveGroup` and curve configuration `ArkCurveConfig`. This gives an
+/// implementation of our `Curve` trait for `ArkGroup<F>` for any `F` that
+/// implements `ark_ec::CurveGroup`, provided an instance of `ArkCurveConfig`
+/// for that curve.
+impl<G: ark_ec::CurveGroup + ArkCurveConfig<G>> Curve for ArkGroup<G>
+where
+    <G as ark_ec::Group>::ScalarField: Serialize,
+{
+    type MultiExpType = GenericMultiExp<Self>;
+    type Scalar = ArkField<<G as ark_ec::Group>::ScalarField>;
+
+    const GROUP_ELEMENT_LENGTH: usize = G::GROUP_ELEMENT_LENGTH;
+    const SCALAR_LENGTH: usize = G::SCALAR_LENGTH;
+
+    fn zero_point() -> Self { ArkGroup(G::zero()) }
+
+    fn one_point() -> Self { ArkGroup(G::generator()) }
+
+    fn is_zero_point(&self) -> bool { self.0.is_zero() }
+
+    fn inverse_point(&self) -> Self { ArkGroup(-self.0) }
+
+    fn double_point(&self) -> Self { ArkGroup(self.0.double()) }
+
+    fn plus_point(&self, other: &Self) -> Self { ArkGroup(self.0 + other.0) }
+
+    fn minus_point(&self, other: &Self) -> Self { ArkGroup(self.0 - other.0) }
+
+    fn mul_by_scalar(&self, scalar: &Self::Scalar) -> Self { ArkGroup(self.0 * scalar.0) }
+
+    fn bytes_to_curve_unchecked<R: byteorder::ReadBytesExt>(b: &mut R) -> anyhow::Result<Self> {
+        // TODO: this is not the most efficient implementation, since there might be
+        // some additional checks during deserialization. However, it seems
+        // there are no unchecked methods available through traits.
+        let res = G::Affine::deserialize_compressed(b).map_err(|e| anyhow!(e))?;
+        Ok(ArkGroup(res.into()))
+    }
+
+    fn generate<R: rand::prelude::Rng>(rng: &mut R) -> Self { ArkGroup(G::rand(rng)) }
+
+    fn generate_scalar<R: rand::prelude::Rng>(rng: &mut R) -> Self::Scalar {
+        <G::ScalarField as ark_ff::UniformRand>::rand(rng).into()
+    }
+
+    fn scalar_from_u64(n: u64) -> Self::Scalar { ArkField(G::ScalarField::from(n)) }
+
+    fn scalar_from_bytes<A: AsRef<[u8]>>(bs: A) -> Self::Scalar {
+        // Traverse at most 4 8-byte chunks, for a total of 256 bits.
+        // The top-most two bits in the last chunk are set to 0.
+        let s = num::integer::div_ceil(Self::Scalar::CAPACITY, 8);
+        let mut fr = Vec::with_capacity(s as usize);
+        for chunk in bs.as_ref().chunks(8).take(s as usize) {
+            let mut v = [0u8; 8];
+            v[..chunk.len()].copy_from_slice(chunk);
+            fr.push(u64::from_le_bytes(v));
+        }
+        // unset two topmost bits in the last read u64.
+        *fr.last_mut().expect("Non empty vector expected") &= !(1u64 << 63 | 1u64 << 62);
+        // fr[3] &= !(1u64 << 63 | 1u64 << 62);
+        <Self::Scalar>::from_repr(&fr)
+            .expect("The scalar with top two bits erased should be valid.")
+    }
```

Comment:
So this is problematic. This `scalar_from_bytes` only makes sense for the BLS curve, but it will not make sense in general. We need to either have this implementation more general, or fix it a specific curve.

Is there no serialization for scalars in ark? I guess there is https://docs.rs/ark-ff/latest/ark_ff/fields/trait.PrimeField.html#method.from_le_bytes_mod_order but that won't be quite the same.

I can't do it now, but we should look into how this method is actually used to see what the best thing to do is.

Reply:
I made the implementation generic. From the documentation of the trait, it is clear what is expected form the implementation.